### PR TITLE
feat(blockifier): add support for keccak syscall

### PR DESCRIPTION
**Pull request by varex83**
This PR adds support for keccak syscall for the native syscall handler.

### CODE REVIEW

**Review by meship-starkware — COMMENTED** [review]
> Reviewed 1 of 3 files at r2, 1 of 2 files at r9, 1 of 1 files at r10.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1546)* status: 3 of 6 files reviewed, 1 unresolved discussion (waiting on @noaov1 and @varex83)

___
*[`crates/blockifier/src/execution/native/syscall_handler.rs` line 233 at r10](https://reviewable.io/reviews/starkware-libs/sequencer/1546#-OAT9QTO5mI00V4LjfiZ:-OAT9QTO5mI00V4Ljfi_:bmbpclv) ([raw file](https://github.com/starkware-libs/sequencer/blob/59867fc7521ce0a11f01760dfcfa450b40f2c53b/crates/blockifier/src/execution/native/syscall_handler.rs#L233)):*
> ```XML
>         )?;
> 
>         const KECCAK_FULL_RATE_IN_WORDS: usize = 17;
> ```

Where did this number come from? 



_Code quote:_
```XML
const KECCAK_FULL_RATE_IN_WORDS: usize = 17;
```

**Review by meship-starkware — COMMENTED** [review]
> Reviewed 1 of 2 files at r9, 4 of 4 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1546)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @noaov1)

**Review by avi-starkware — APPROVED** [review]
> Reviewed 1 of 1 files at r10, 4 of 4 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1546)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @noaov1)

**Review by meship-starkware — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1546)* status: all files reviewed, 1 unresolved discussion (waiting on @varex83)

**Review by meship-starkware — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1546)* status: all files reviewed, 1 unresolved discussion (waiting on @varex83)

**Review by meship-starkware — APPROVED** [review]
> Reviewed 5 of 5 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1546)* status: all files reviewed, 1 unresolved discussion (waiting on @varex83)

**Review by meship-starkware — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1546#-:-OBQo1G95SkP4bbAnZes:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @varex83)

**Review by meship-starkware — APPROVED** [review]
> Reviewed 2 of 2 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1546)* status: all files reviewed, 1 unresolved discussion (waiting on @varex83)

**Review by meship-starkware — APPROVED** [review]
You have some conflicts please rebase 

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1546#-:-OBVsOm79Es6DtCHhQgp:b-rbbwg1)* status: all files reviewed, 1 unresolved discussion (waiting on @varex83)

**Review by meship-starkware — APPROVED** [review]
> Reviewed 2 of 2 files at r14, all commit messages.
> Dismissed @graphite-app[bot] from a discussion.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1546)* status: all files reviewed, 2 unresolved discussions (waiting on @varex83 and @Yoni-Starkware)

**Review by rodrigo-pino — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1546)* status: all files reviewed, 2 unresolved discussions (waiting on @noaov1, @varex83, and @Yoni-Starkware)

___
*[`crates/blockifier/src/execution/native/syscall_handler.rs` line 226 at r11](https://reviewable.io/reviews/starkware-libs/sequencer/1546#-OB117yw6Jz3B814dIOe:-OB_64X2Cid3mOsjQ7EH:bdb4t8w) ([raw file](https://github.com/starkware-libs/sequencer/blob/c9d0ba18cbfbe867a0e2acfbae20fdd2aa138918/crates/blockifier/src/execution/native/syscall_handler.rs#L226)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

In the CASM syscall implementation, we receive input as a vector of felts. If any of the felts don't fit into a `u64`, we return an (unprovable) error. I'm curious about how Cairo native would handle this situation (is it even possible?).
</blockquote></details>

 We have no control over it, we would have to ask the Native team for certainty but my guess is that it would throw an overflow error (if it was possible for it to happen at first). Since one of the pros of Sierra is that there is type information present, and if execution was correct until this point (no overflow errors whatsoever) my guess is that those values won't be incorrect.

**Review by meship-starkware — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1546)* status: all files reviewed, 1 unresolved discussion (waiting on @varex83 and @Yoni-Starkware)

File: crates/blockifier/src/execution/native/syscall_handler.rs [hunk 1]
```diff
@@ -245,8 +249,51 @@ impl<'state> StarknetSyscallHandler for &mut NativeSyscallHandler<'state> {
...
+        }
+        *remaining_gas -= gas_cost;
+
+        let mut state = [0u64; 25];
+        for chunk in input.chunks(KECCAK_FULL_RATE_IN_WORDS) {
+            for (i, val) in chunk.iter().enumerate() {
+                state[i] ^= val;
+            }
+            keccak::f1600(&mut state)
+        }
```
[the remark this answers is not in the document]
Comment (avi-starkware) [reply crates/blockifier/src/execution/native/syscall_handler.rs:291]: @ilyalesokhin-starkware 
Is this comment correct?

**Review by meship-starkware — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1546)* status: all files reviewed, 1 unresolved discussion (waiting on @ilyalesokhin-starkware and @Yoni-Starkware)

**Review by Yoni-Starkware — APPROVED** [review]
:lgtm:

> Reviewed 1 of 2 files at r13, 2 of 2 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1546#-:-OBeL5edEPmx5uvnnOg0:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @ilyalesokhin-starkware)

File: crates/blockifier/src/execution/native/syscall_handler.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ilyalesokhin-starkware) [reply crates/blockifier/src/execution/native/syscall_handler.rs:291]: The syscall computes a keccak round; the caller is responsible for the padding.

[the remark this answers is not in the document]
Comment (Yoni-Starkware) [reply crates/blockifier/src/execution/native/syscall_handler.rs:291]: Thanks @ilyalesokhin-starkware

**Review by Yoni-Starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1546)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @varex83)

**Comment by ilyalesokhin-starkware** [comment]
*[`crates/blockifier/src/execution/native/syscall_handler.rs` line 291 at r14](https://reviewable.io/reviews/starkware-libs/sequencer/1546#gh-1839758018:-OBeODDrDvrNCsjnqSY0:br67ovx) ([raw file](https://github.com/starkware-libs/sequencer/blob/6479b4408887b97ef78f58bc0f297b3916c0f1ca/crates/blockifier/src/execution/native/syscall_handler.rs#L291)):*
<details><summary><i>Previously, Yoni-Starkware (Yoni) wrote…</i></summary><blockquote>

Thanks @ilyalesokhin-starkware
</blockquote></details>

See:
https://github.com/starkware-libs/cairo/blob/88cefed325bd1039f80641b55220a0c84dd02ad8/corelib/src/starknet/syscalls.cairo#L96

**Pull request merged**